src/persist-client/src/read.rs

```diff
@@ -76,43 +73,143 @@ pub(crate) enum HollowBatchReaderMetadata<T> {
+            Issued { seqno, .. } | Consumed { seqno, .. } => Some(*seqno),
+            Completed => None,
+        }
+    }
+}
+
 /// A token representing one read batch.
 ///
 /// This may be exchanged (including over the network). It is tradeable via
-/// [ReadHandle::fetch_batch] for the resulting data stored in the batch.
-#[derive(Debug, Clone, Serialize, Deserialize)]
-#[serde(bound(
-    serialize = "T: Timestamp + Codec64",
-    deserialize = "T: Timestamp + Codec64"
-))]
-#[serde(
-    into = "SerdeReaderEnrichedHollowBatch",
-    from = "SerdeReaderEnrichedHollowBatch"
-)]
-pub struct ReaderEnrichedHollowBatch<T> {
+/// `BatchFetcher::fetch_batch` for the resulting data stored in the batch.
+///
+/// # Panics
+/// `LeasedBatch` panics when dropped unless a very strict set of invariants are
+/// held:
+///
+/// `LeasedBatch` may only be dropped if it:
+/// - Is being converted to `SerdeLeasedBatch`.
+/// - Has been returned to its issuer via `Self::return_lease`.
+///
+/// In any other circumstance, dropping `LeasedBatch` panics.
+#[derive(Debug)]
+pub struct LeasedBatch<T>
+where
+    T: Timestamp + Codec64,
+{
     pub(crate) shard_id: ShardId,
-    pub(crate) reader_metadata: HollowBatchReaderMetadata<T>,
+    pub(crate) reader_id: ReaderId,
+    pub(crate) metadata: LeasedBatchMetadata<T>,
     pub(crate) batch: HollowBatch<T>,
+    /// The `SeqNo` from which this batch originated; we track this value as
+    /// long as necessary to ensure the `SeqNo` isn't garbage collected while a
+    /// read still depends on it.
+    pub(crate) leased_seqno: LeaseLifeCycle,
 }
 
-impl<T> ReaderEnrichedHollowBatch<T>
+impl<T> LeasedBatch<T>
 where
-    T: Timestamp + Lattice + Codec64,
+    T: Timestamp + Codec64,
 {
+    pub(crate) fn new(
+        shard_id: ShardId,
+        reader_id: ReaderId,
+        metadata: LeasedBatchMetadata<T>,
+        batch: HollowBatch<T>,
+        leased_seqno: LeaseLifeCycle,
+    ) -> LeasedBatch<T> {
+        LeasedBatch {
+            shard_id,
+            reader_id,
+            metadata,
+            batch,
+            leased_seqno,
+        }
+    }
+
     /// Signals whether or not `self` should downgrade the `Capability` its
     /// presented alongside.
     pub fn generate_progress(&self) -> Option<Antichain<T>> {
-        match self.reader_metadata {
-            HollowBatchReaderMetadata::Listen { .. } => Some(self.batch.desc.upper().clone()),
-            HollowBatchReaderMetadata::Snapshot { .. } => None,
+        match self.metadata {
+            LeasedBatchMetadata::Listen { .. } => Some(self.batch.desc.upper().clone()),
+            LeasedBatchMetadata::Snapshot { .. } => None,
         }
     }
+
+    /// Identifies that the batch has been consumed. After calling this
+    /// function, the batch must either be returned to its issuer, or promise a
+    /// return.
+    ///
+    /// # Panics
+    /// - If lease has already been consumed
+    fn consume_lease(&mut self) {
```

Comment:
I'm not sure I follow why there needs to be 3 states?

in my mental model, we just need 2: leased and returned. the latter is just what `process_returned_leased_batch` uses to keep drop from panic'ing. this would mean we could replace LeaseLifeCycle with an `Option<SeqNo>`

very possible I'm missing something though, y'all have been thinking about this much more than I have

Reply:
This was just over-engineered during dev and seems like a good idea to simplify